Sources/SwiftGRPCNIO/ClientCalls/BaseClientCall.swift

```diff
@@ -0,0 +1,177 @@
+    self.sendRequestHead(requestHead)
+  }
+}
+
+extension BaseClientCall: ClientCall {
+  /// HTTP/2 stream associated with this call.
+  public var subchannel: EventLoopFuture<Channel> {
+    return self.streamPromise.futureResult
+  }
+
+  /// Initial metadata returned from the server.
+  public var initialMetadata: EventLoopFuture<HTTPHeaders> {
+    return self.clientChannelHandler.initialMetadataPromise.futureResult
+  }
+
+  /// Status of this call which may originate from the server or client.
+  ///
+  /// Note: despite `GRPCStatus` being an `Error`, the value will be delievered as a __success__
+  /// result even if the status represents a __negative__ outcome.
+  public var status: EventLoopFuture<GRPCStatus> {
+    return self.clientChannelHandler.statusPromise.futureResult
+  }
+
+  /// Cancel the current call.
+  ///
+  /// Closes the HTTP/2 stream once it becomes available. Additional writes to the channel will be ignored.
+  /// Any unfulfilled promises will be failed with a cancelled status (excepting `status` which will be
+  /// succeeded, if not already succeeded).
+  public func cancel() {
+    self.client.channel.eventLoop.execute {
+      self.subchannel.whenSuccess { channel in
+        channel.close(mode: .all, promise: nil)
+      }
+    }
+  }
+}
+
+extension BaseClientCall {
+  /// Creates and configures an HTTP/2 stream channel. `subchannel` will contain the stream channel when it is created.
+  internal func createStreamChannel() {
+    self.client.channel.eventLoop.execute {
+      self.client.multiplexer.createStreamChannel(promise: self.streamPromise) { (subchannel, streamID) -> EventLoopFuture<Void> in
+        subchannel.pipeline.addHandlers([HTTP2ToHTTP1ClientCodec(streamID: streamID, httpProtocol: .http),
+                                         HTTP1ToRawGRPCClientCodec(),
+                                         GRPCClientCodec<RequestMessage, ResponseMessage>(),
+                                         self.clientChannelHandler],
+                                        first: false)
+      }
+    }
+  }
+
+  /// Send the request head once `subchannel` becomes available.
+  internal func sendRequestHead(_ requestHead: HTTPRequestHead) {
+    self.subchannel.whenSuccess { channel in
+      channel.write(GRPCClientRequestPart<RequestMessage>.head(requestHead), promise: nil)
+    }
+  }
+
+  /// Send the given request once `subchannel` becomes available.
+  internal func sendRequest(_ request: RequestMessage) {
+    self.subchannel.whenSuccess { channel in
+      channel.write(GRPCClientRequestPart<RequestMessage>.message(request), promise: nil)
+    }
+  }
+
+  /// Send `end` once `subchannel` becomes available.
+  internal func sendEnd() {
+    self.subchannel.whenSuccess { channel in
+      channel.writeAndFlush(GRPCClientRequestPart<RequestMessage>.end, promise: nil)
```

Comment:
We flush after sending `.end`, anyway, but I guess it won't hurt to `writeANDFLUSH` here, anyway.